DM-15471: Control disassembly at butler config level not storage class

timj opened the pull request:
Previously StorageClasses defined how to disassembly, but if we want per-datasetType control over disassembly we have to move them into their own configuration area.

timj committed:
Fix typo in storage class name

timj committed:
Rename composites.py to assembler.py

Makes more sense for the default assembler to be an assembler.

timj committed:
Add logging to report when a config file is read

timj committed:
Control composite disassembly by butler config

No longer control it at storage classes definition.

timj committed:
Enable calexp reassembly

timj committed:
Remove unnecessary logging of ephemeral status

TallJimbo commented on config/datastores/posixDatastore.yaml:
```diff
@@ -35,3 +35,5 @@ datastore:
     Config: lsst.daf.butler.formatters.pexConfigFormatter.PexConfigFormatter
     Packages: lsst.daf.butler.formatters.pickleFormatter.PickleFormatter
+    PropertyList: lsst.daf.butler.formatters.pickleFormatter.PickleFormatter
+    PropertySet: lsst.daf.butler.formatters.pickleFormatter.PickleFormatter
```
Probably not an issue for this ticket, but we should have and use a YAML formatter for these.

timj replied:
I need to work out what is going on with DM-14503 -- in that ticket YAML support was added to PropertyList by adding it directly to daf_persistence (lsst/daf_persistence#94). Should we add YAML support natively to daf_base itself so that `yaml.dump()` works independently of daf_persistence? Is the issue that we didn't want daf_base to depend on yaml (unlike pickle which is added natively)? @ktlim ?

TallJimbo replied:
Oh, I hadn't realized the YAML I/O was not in daf_base.  I'd definitely like to see it moved there, though I wouldn't object to having daf_butler depend on daf_persistence temporarily.

TallJimbo commented on config/composites.yaml:
```diff
@@ -0,0 +1,17 @@
+    ExposureI: False
+    ExposureF: False
+    ExposureCompositeF: True
+    ExposureCompositeF: True
```
Is it possible to set these options for the `Exposure` "base" StorageClass and have it propagate to "derived" StorageClasses?

timj replied:
StorageClass doesn't have a hierarchy in python land so that's not possible. We could change that by changing the way we define storage classes to make them explicitly state a parent rather than using YAML substitution.

TallJimbo replied:
Ok, let's just hope [DM-15238](https://jira.lsstcorp.org/browse/DM-15238) makes this a moot point.

andy-slac replied:
Unrelated - there are two `ExposureCompositeF`, should one of them be `ExposureCompositeI`?

timj replied:
I don't think true python inheritance of storage classes is all that hard if we need it.

TallJimbo commented on python/lsst/daf/butler/core/composites.py:
```diff
@@ -0,0 +1,129 @@
+        if isStr:
+            disassemble = False
+        else:
+            disassemble = self.config["default"]
```
I don't understand why the default should be different for `str` vs. `DatasetType` or `StorageClass` instances.

timj replied:
For DatasetType and Storageclass you know for a fact that the requested item is a composite and therefore the default matters. For a `str` you have no idea whether you are being given a composite and returning True for disassembly could be very dangerous.

One option is to drop `str` support completely.

TallJimbo replied:
> One option is to drop str support completely.

If it doesn't make other code uglier, I'm all for it.  Seems like this is a pretty internal API so I'd lean towards simplifying it even if that reduces flexibility.

TallJimbo commented on config/storageClasses.yaml:
```diff
@@ -50,7 +52,7 @@ storageClasses:
       apCorrMap: TablePersistableApCorr
       coaddInputs: TablePersistableCoaddInputs
-      metadata: PropertyList
+      metadata: PropertySet
```
Huh.  Just an FYI: I was under the impression that the thing returned by `Exposure.getMetadata()` was always a `PropertyList`, but it turns out that's because it's implemented that way and pybind11 always downcasts to the most-derived type; the C++ does say `PropertySet`.  I'm inclined to try to get that changed in afw, because I think we really do depend on it being a `PropertyList` in various places, so this will probably get reverted eventually.

TallJimbo reviewed (approved):
Looks good!  A few minor comments (on individual commits), but nothing I actually object to.

timj committed:
Remove assembler from DatasetRef

timj committed:
Fix stringification of StorageClass

Empty component list was not working properly and we always
have a dict so can always use `keys()`.

timj committed:
Fix typo in ExposureComposite composites file
